subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rules/ComponentMetadataRulesIntegrationTest.groovy

```diff
@@ -82,6 +82,109 @@ dependencies {
         succeeds 'resolve'
     }
 
+    @RequiredFeatures([
+        @RequiredFeature(feature = GradleMetadataResolveRunner.REPOSITORY_TYPE, value = "maven"),
+        @RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value = "false")
+    ])
```

Comment:
I wonder if we should remove the `@RequiredFeatures` here to test also the other cases for which we have no test coverage so far.
That would mean that the `assert`s further down would be conditionally. That is, if there is Ivy or Gradle metadata, the fields will be `null` (because the classifier on the dependency is not published).
You can use `useIvy()` and `isGradleMetadataPublished()` to test.

 

There is also a way to publish `classifiers` in GMM as a compatibility layer for Maven. I am not sure if we need to pay specific attention to that, but it would be interesting to know how that behaves: https://github.com/gradle/gradle/blob/master/subprojects/docs/src/docs/design/gradle-module-metadata-latest-specification.md#thirdpartycompatibility-value